server/api-util/idToken.js

```diff
@@ -0,0 +1,82 @@
+const crypto = require('crypto');
+const { default: fromKeyLike } = require('jose/jwk/from_key_like');
+const { default: SignJWT } = require('jose/jwt/sign');
+
+const radix = 10;
+const PORT = parseInt(process.env.REACT_APP_DEV_API_SERVER_PORT, radix);
+const rootUrl = process.env.REACT_APP_CANONICAL_ROOT_URL;
+const useDevApiServer = process.env.NODE_ENV === 'development' && !!PORT;
+
+const issuerUrl = useDevApiServer ? `http://localhost:${PORT}/api` : `${rootUrl}/api`;
+
+const clientId = process.env.CUSTOM_OIDC_CLIENT_ID;
+
+// These keys are used for signing the ID token (JWT)
+// When you store them to environment variables you should replace
+// any line brakes with '\n'.
+// You should also make sure that the key size is big enough.
+const rsaSecretKey = process.env.RSA_SECRET_KEY;
+const rsaPublicKey = process.env.RSA_PUBLIC_KEY;
+
+/**
+ * Gets user information and creates the signed jwt for id token.
+ *
+ * @param {Object} user
+ *
+ * @return {Promise} idToken
+ */
+exports.createIdToken = user => {
+  if (!user) {
+    console.error('Missing user information!');
+    return;
+  }
+
+  if (!(rsaSecretKey && rsaPublicKey)) {
+    console.error('Missing RSA keys!');
+    return;
+  }
+
+  // We use jose library which requires the RSA key
+  // to be KeyLike format:
+  // https://github.com/panva/jose/blob/master/docs/types/_types_d_.keylike.md
+  const privateKey = crypto.createPrivateKey(rsaSecretKey);
+
+  const { userId, profile } = user;
+
+  const jwt = new SignJWT({
+    given_name: profile.firstName,
+    family_name: profile.lastName,
+    email: profile.email,
+    email_verified: profile.emailVerified,
+  })
+    .setProtectedHeader({ alg: 'RS256' })
+    .setIssuedAt()
+    .setIssuer(issuerUrl)
+    .setSubject(userId)
+    .setAudience(clientId)
+    .setExpirationTime('1h')
+    .sign(privateKey);
```

Comment:
Did we have the `kid` value in the token header? I might have mentioned that it's not compulsory but maybe we could add it by default to the ID token. Even though it's one more random value for the user to decide, enforcing the `kid` value in the token and in JWK could be useful as key caching in the Flex API relies heavily on it.

Reply:
Yes, we had but then I removed it to test if it works without it (and it did). I didn't want to leave it there with any hardcoded value. But I think it makes sense to have it in the example so it there again. Now the key id is also read from environment variables so it can be basically any string user decides. I guess this is ok?